LayoutKit/Views/ReloadableViewUpdateManager.swift

```diff
@@ -0,0 +1,138 @@
+// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+
+import Foundation
+
+/// Delegate for a `ReloadableViewUpdateManager`.
+protocol ReloadableViewUpdateManagerDelegate: class {
+    weak var reloadableView: ReloadableView? { get }
+    var currentArrangement: [Section<[LayoutArrangement]>] { get set }
+}
+
+/// Manages updating a `ReloadableView` and its backing data.
+/// Updates may be applied incrementally, or in a batch.
+class ReloadableViewUpdateManager {
+    private weak var delegate: ReloadableViewUpdateManagerDelegate?
+    private weak var operation: NSOperation?
+
+    static func make(delegate delegate: ReloadableViewUpdateManagerDelegate,
+                              operation: NSOperation,
+                              incremental: Bool) -> ReloadableViewUpdateManager {
+        return incremental ?
+            IncrementalUpdateManager(delegate: delegate, operation: operation) :
+            BatchUpdateManager(delegate: delegate, operation: operation)
+    }
+
+    private init(delegate: ReloadableViewUpdateManagerDelegate, operation: NSOperation) {
+        self.delegate = delegate
+        self.operation = operation
+    }
+
+    func apply(partialArrangement arrangement: [Section<[LayoutArrangement]>], insertedIndexPath: NSIndexPath) {
+        // implemented by subclass
+    }
+
+    func apply(finalArrangement arrangement: [Section<[LayoutArrangement]>], batchUpdates: BatchUpdates?, completion: (() -> Void)?) {
+        // implemented by subclass
+    }
+
+    final func updateReloadableView(waitUntilFinished waitUntilFinished: Bool, updates: (reloadableView: ReloadableView) -> Void) {
+        let mainOperation = NSBlockOperation()
+        mainOperation.addExecutionBlock {
+            let operationCancelled = self.operation?.cancelled ?? true
+            guard !operationCancelled, let reloadableView = self.delegate?.reloadableView else {
+                return
+            }
+            updates(reloadableView: reloadableView)
+        }
+        NSOperationQueue.mainQueue().addOperations([mainOperation], waitUntilFinished: waitUntilFinished)
+    }
+}
+
+/// Updates the `ReloadableView` incrementally as items are inserted.
+private class IncrementalUpdateManager: ReloadableViewUpdateManager {
+
+    private var pendingInsertedIndexPaths = [NSIndexPath]()
+
+    override func apply(partialArrangement arrangement: [Section<[LayoutArrangement]>], insertedIndexPath: NSIndexPath) {
+        updateReloadableView(waitUntilFinished: false) { (reloadableView: ReloadableView) in
+            self.pendingInsertedIndexPaths.append(insertedIndexPath)
+
+            // Don't modify the data while the view is moving.
```

Comment:
Is this true of all reloads when moving? Wondering if this library should handle this or leave it to the user. Because a lot of features require things to edit immediatley?